tests/test_timer.cpp

```diff
@@ -32,164 +33,555 @@ using namespace testing;
-    ASSERT_FALSE(timedout.load()) << "Should not have timed out yet";
+
+    {
+        // Wait for the the timer to be "running" in the mocked wait_for()
+        std::unique_lock<std::mutex> test_lock(cv_m);
+        cv.wait(test_lock, [this] { return timer_running.load(); });
+    }
 
     t.pause();
-    std::this_thread::sleep_for(5ms);
+
+    {
+        // Wait for the the timer to be "paused" in the mocked wait()
+        std::unique_lock<std::mutex> test_lock(cv_m);
+        cv.wait(test_lock, [this] { return timer_paused.load(); });
+    }
 
     ASSERT_FALSE(timedout.load()) << "Should not have timed out";
 }
 
-TEST_F(Timer, resumes)
+TEST_F(TestTimer, resumes)
 {
-    const std::chrono::milliseconds timeout(10);
-
-    mpu::Timer t{duration_cast<seconds>(timeout), [this]() {
+    EXPECT_CALL(*mock_timer_sync_funcs, wait_for(_, _, _))
+        .WillOnce([this](auto&, auto& timer_lock, auto&) {
+            std::unique_lock<std::mutex> test_lock(cv_m);
+
+            // Unlock the Timer's internal lock
+            timer_lock.unlock();
+
+            // Notify the test thread that the timer is running
+            timer_running.store(true);
+            cv.notify_all();
+
+            // Wait to be notified by the mock notify_all()
+            cv.wait(test_lock, [this] { return test_notify_called.load(); });
+            test_notify_called.store(false);
+
+            // Re-lock the Timer's internal lock before returning
+            timer_lock.lock();
+
+            return std::cv_status::no_timeout;
+        })
+        .WillOnce([](auto&...) { return std::cv_status::timeout; });
+
+    EXPECT_CALL(*mock_timer_sync_funcs, notify_all(_))
+        .Times(4)
+        .WillOnce(Return())             // For the stop() in start()
+        .WillRepeatedly([this](auto&) { // Handles the 3 other notify_all() calls
+            {
+                std::unique_lock<std::mutex> test_lock(cv_m);
+                test_notify_called.store(true);
+            }
+            // Notify the mocked wait() and wait_for()
+            cv.notify_all();
+        });
+
+    EXPECT_CALL(*mock_timer_sync_funcs, wait(_, _)).WillOnce([this](auto&, auto& timer_lock) {
+        std::unique_lock<std::mutex> test_lock(cv_m);
+
+        timer_paused.store(true);
+        cv.notify_all();
+
+        // Unlock the Timer's internal lock
+        timer_lock.unlock();
+
+        // Waits until resume is called()
+        cv.wait(test_lock, [this] { return test_notify_called.load(); });
+        test_notify_called.store(false);
+
+        // Re-lock the Timer's internal lock before returning
+        timer_lock.lock();
+    });
+
+    mpu::Timer t{default_timeout, [this]() {
                      {
-                         std::lock_guard<std::mutex> lk{cv_m};
+                         std::lock_guard<std::mutex> test_lock{cv_m};
                          timedout.store(true);
                      }
                      cv.notify_all();
                  }};
```

Comment:
This too we should be able to avoid repeating. Maybe make the `Timer` obj an attribute of `TimerTest`? I know there are a few variations, but a generic form should be enough, I think. Perhaps dropping the `timedout` boolean and replacing with a `timedout_count` to fit `Timer.restarts too`.

Reply:
Good ideas, I'll look into this as well.

Reply:
Ok, done.